{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This script works to find to which range corresponds a sequence from the reference genome of the pangenome database built with phgv2, and check how many alleles does it have.\n",
    "\n",
    "It needs the merged hvcf of the pangenome database, and then the coordinates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def AmIaNotebook():\n",
    "    try:\n",
    "        shell = get_ipython().__class__.__name__\n",
    "        if shell == 'ZMQInteractiveShell':\n",
    "            return True  # Jupyter notebook or qtconsole\n",
    "        elif shell == 'TerminalInteractiveShell':\n",
    "            return False  # Terminal running IPython\n",
    "        else:\n",
    "            return False  # Other type of environment\n",
    "    except NameError:\n",
    "        return False  # Not in an interactive environment\n",
    "\n",
    "#Check if the name of this file ends with .py or with .ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def SetupCheckHaplo(hvcf, reference_fasta):\n",
    "    \"\"\"\n",
    "    If there are coordintates, use them.\n",
    "    If not,they will be asked\n",
    "    \"\"\"\n",
    "\n",
    "    #If you leave this empty it will be asked later\n",
    "    #Coordinates:\n",
    "    start = None\n",
    "    end = None\n",
    "    chromosome = None\n",
    "\n",
    "    if not hvcf or reference_fasta:\n",
    "        raise ValueError(\"You must provide the path to the vcf and the reference\")\n",
    "        exit(1)\n",
    "\n",
    "    # Check if start and end are provided, if not, prompt the user\n",
    "    if start is None and end is None and chromosome is None:\n",
    "        start = int(input(\"Start: \"))\n",
    "        end = int(input(\"End: \"))\n",
    "        chromosome = input(\"Chromosome: (enter only the number)\")\n",
    "    else:\n",
    "        raise ValueError(\"You must provide either all or none of the coordinates\")\n",
    "\n",
    "    return start, end, chromosome\n",
    "\n",
    "    \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def DefineRange(hvcf, reference_fasta):\n",
    "    \n",
    "    start, end, chromosome = SetupCheckHaplo(hvcf, reference_fasta)\n",
    "    print(\"Start: \", start)\n",
    "    print(\"End: \", end)\n",
    "    chromosome = \"chr\" + chromosome\n",
    "    print(\"Chromosome: \", chromosome, \"\\n\")\n",
    "\n",
    "    keys = []\n",
    "\n",
    "    with open(hvcf, 'r') as f:\n",
    "        for line in f:\n",
    "            if line.startswith(\"#\"):\n",
    "                continue\n",
    "            else:\n",
    "                line = line.split(\"\\t\")\n",
    "                ref_start = line[1]\n",
    "                #capture the element of the list starting with \"END\"\n",
    "                ref_end = [i for i in line if i.startswith(\"END\")]\n",
    "                ref_end = ref_end[0].split(\"=\")[1]\n",
    "\n",
    "                if int(ref_start) <= start and int(ref_end) >= end:\n",
    "                    if line[0].startswith(chromosome):\n",
    "                        keys = line[4].split(\",\")\n",
    "                        keys = [i.strip(\"<>\") for i in keys]\n",
    "\n",
    "    if keys == []:\n",
    "        print(\"No range found containing the coordinates\")\n",
    "        return None\n",
    "    else:\n",
    "\n",
    "        print (f\"there are {len(keys)} keys of ranges containing the coordinates:\")\n",
    "        print(keys)\n",
    "        print(\"\\n\")\n",
    "        return keys\n",
    "\n",
    "                    \n",
    "                    \n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def GetCoordinates(hvcf, reference_fasta):\n",
    "\n",
    "    keys = DefineRange(hvcf, reference_fasta)\n",
    "\n",
    "    if keys is None:\n",
    "        pass\n",
    "\n",
    "    else:\n",
    "        with open(hvcf, 'r') as f:\n",
    "            for line in f:\n",
    "                if not line.startswith(\"#\"):\n",
    "                    continue\n",
    "                else:\n",
    "                    for key in keys:\n",
    "                        if key in line:\n",
    "                            line = line.split(\",\")\n",
    "                            samplename = line[3]\n",
    "                            samplename = samplename.split(\"=\")[1]\n",
    "                            region = line[4]\n",
    "                            region = region.split(\"=\")[1]\n",
    "                            ref_range = line[6].split(\"=\")[1]\n",
    "                            ref_range = ref_range[:-1] #remove the last character \">\"\n",
    "\n",
    "                            print(f\"Line: {samplename}\\tRegion: {region}\\tReference range:{ref_range} \")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def ___main___():\n",
    "\n",
    "    \"\"\"\n",
    "    This script will take a vcf file and a reference fasta file and will return the coordinates of the ranges that contain the coordinates provided by the user.\n",
    "    Provide the path to the vcf file and the reference fasta file.\n",
    "    If executing by command line, you can provide the coordinates as arguments. If not, you will be asked for them.\n",
    "    \"\"\"\n",
    "\n",
    "    import os\n",
    "    import pandas as pd\n",
    "    import argparse\n",
    "    import sys\n",
    "\n",
    "\n",
    "    if AmIaNotebook() == False:\n",
    "        parser = argparse.ArgumentParser(description=___main___.__doc__)\n",
    "        parser.add_argument('--hvcf', \"-hf\", type=str, help='The hvcf file', required=True)\n",
    "        parser.add_argument('--reference_fasta', \"-ref\", type=str, help='The reference fasta file', required=True)\n",
    "        parser.add_argument('--start', \"-s\", type=int, help='The start coordinate')\n",
    "        parser.add_argument('--end', \"-e\", type=int, help='The end coordinate')\n",
    "        parser.add_argument('--chromosome', \"-c\", type=str, help='The chromosome')\n",
    "        args = parser.parse_args()\n",
    "\n",
    "        start = args.start\n",
    "        end = args.end\n",
    "        chromosome = args.chromosome\n",
    "        hvcf = args.hvcf\n",
    "        reference_fasta = args.reference_fasta\n",
    "\n",
    "    else:\n",
    "        hvcf = \"/scratch/PHG/output/vcf_files/merged_hvcfs_19092024.h.vcf\"\n",
    "        reference_fasta = \"/scratch/PHG/data/prepared_genomes/MorexV3.fa\"\n",
    "\n",
    "    GetCoordinates(hvcf, reference_fasta)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Start:  1000\n",
      "End:  10000\n",
      "Chromosome:  chr4 \n",
      "\n",
      "No range found containing the coordinates\n"
     ]
    }
   ],
   "source": [
    "if __name__ == \"__main__\":\n",
    "    try:\n",
    "        ___main___()    \n",
    "    except KeyboardInterrupt:\n",
    "        print(\"Script interrupted by user. Exiting...\")\n",
    "        sys.exit(0)\n",
    "        raise KeyboardInterrupt"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pygenometracks",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
