{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d58f7f84",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# This script works to find to which range corresponds a sequence from the reference genome of the pangenome database built with phgv2, and check how many alleles does it have.\n",
    "# \n",
    "# It needs the merged hvcf of the pangenome database, and then the coordinates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "659716a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "def AmIaNotebook():\n",
    "    try:\n",
    "        shell = get_ipython().__class__.__name__\n",
    "        if shell == 'ZMQInteractiveShell':\n",
    "            return True  # Jupyter notebook or qtconsole\n",
    "        elif shell == 'TerminalInteractiveShell':\n",
    "            return False  # Terminal running IPython\n",
    "        else:\n",
    "            return False  # Other type of environment\n",
    "    except NameError:\n",
    "        return False  # Not in an interactive environment\n",
    "\n",
    "#Check if the name of this file ends with .py or with .\n",
    "\n",
    "\n",
    "def DefineRange(hvcf, reference_fasta, start, end, chromosome):\n",
    "    \n",
    "    print(\"Start: \", start)\n",
    "    print(\"End: \", end)\n",
    "    chromosome = \"chr\" + chromosome\n",
    "    print(\"Chromosome: \", chromosome, \"\\n\")\n",
    "\n",
    "\n",
    "\n",
    "    keys = []\n",
    "    new_keys = []\n",
    "\n",
    "    with open(hvcf, 'r') as f:\n",
    "        for line in f:\n",
    "            if line.startswith(\"#\"):\n",
    "                continue\n",
    "            else:\n",
    "                line = line.split(\"\\t\")\n",
    "                ref_start = line[1]\n",
    "                #capture the element of the list starting with \"END\"\n",
    "                ref_end = [i for i in line if i.startswith(\"END\")]\n",
    "                ref_end = ref_end[0].split(\"=\")[1]\n",
    "\n",
    "                #Check all ranges that overlap the coordinates\n",
    "                if int(ref_start) >= start and int(ref_end) <= end:\n",
    "\n",
    "                    if line[0].startswith(chromosome):\n",
    "                        new_keys = line[4].split(\",\")\n",
    "                        new_keys = [i.strip(\"<>\") for i in new_keys]\n",
    "                        keys.extend(new_keys)\n",
    "\n",
    "                #check if a single range contains the coordinates\n",
    "                elif int(ref_start) <= start and int(ref_end) >= end:\n",
    "                    if line[0].startswith(chromosome):\n",
    "                        new_keys = line[4].split(\",\")\n",
    "                        new_keys = [i.strip(\"<>\") for i in new_keys]\n",
    "                        keys.extend(new_keys)\n",
    "\n",
    "\n",
    "    if keys == []:\n",
    "        print(\"No range found containing the coordinates\")\n",
    "        return None\n",
    "    else:\n",
    "\n",
    "        print (f\"there are {len(keys)} keys of ranges containing the coordinates:\")\n",
    "        print(keys)\n",
    "        print(\"\\n\")\n",
    "        return keys"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c4f18a80",
   "metadata": {
    "lines_to_next_cell": 2
   },
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "def GetCoordinates(hvcf, reference_fasta, start, end, chromosome):\n",
    "\n",
    "    keys = DefineRange(hvcf, reference_fasta, start, end, chromosome)\n",
    "\n",
    "    if keys is None:\n",
    "        pass\n",
    "\n",
    "    else:\n",
    "        with open(hvcf, 'r') as f:\n",
    "            for line in f:\n",
    "                if not line.startswith(\"#\"):\n",
    "                    continue\n",
    "                else:\n",
    "                    for key in keys:\n",
    "                        if key in line:\n",
    "                            line = line.split(\",\")\n",
    "                            samplename = line[3]\n",
    "                            samplename = samplename.split(\"=\")[1]\n",
    "                            region = line[4]\n",
    "                            region = region.split(\"=\")[1]\n",
    "                            ref_range = line[6].split(\"=\")[1]\n",
    "                            ref_range = ref_range[:-1] #remove the last character \">\"\n",
    "\n",
    "                            print(f\"Line: {samplename}\\tRegion: {region}\\tReference range:{ref_range} \")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fe1b0d73",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "def main():\n",
    "\n",
    "    \"\"\"\n",
    "    This script will take a vcf file and a reference fasta file and will return the coordinates of the ranges that contain the coordinates provided by the user.\n",
    "    Provide the path to the vcf file and the reference fasta file.\n",
    "    If executing by command line, you can provide the coordinates as arguments. If not, you will be asked for them.\n",
    "    \"\"\"\n",
    "\n",
    "    import os\n",
    "    import pandas as pd\n",
    "    import argparse\n",
    "    import sys\n",
    "\n",
    "\n",
    "    if AmIaNotebook() == False:\n",
    "        parser = argparse.ArgumentParser(description=main.__doc__)\n",
    "        parser.add_argument('--hvcf', \"-hf\", type=str, help='The hvcf file', required=True)\n",
    "        parser.add_argument('--reference-fasta', \"-ref\", type=str, help='The reference fasta file', required=True)\n",
    "        parser.add_argument('--start', \"-s\", type=int, help='The start coordinate')\n",
    "        parser.add_argument('--end', \"-e\", type=int, help='The end coordinate')\n",
    "        parser.add_argument('--chromosome', \"-c\", type=str, help='The chromosome')\n",
    "        args = parser.parse_args()\n",
    "\n",
    "        start = args.start\n",
    "        end = args.end\n",
    "        chromosome = args.chromosome\n",
    "        hvcf = args.hvcf\n",
    "        reference_fasta = args.reference_fasta\n",
    "\n",
    "    else:\n",
    "        hvcf = \"/scratch/PHG/output/vcf_files/merged_hvcfs_19092024.h.vcf\"\n",
    "        reference_fasta = \"/scratch/PHG/data/prepared_genomes/MorexV3.fa\"\n",
    "        start = None        #Add the coordinates here if you want to provide them\n",
    "        end = None          #Add the coordinates here if you want to provide them\n",
    "        chromosome = None   #Add the chromosome here if you want to provide it\n",
    "\n",
    "    if start is None and end is None and chromosome is None:\n",
    "        start = int(input(\"Start: \"))\n",
    "        end = int(input(\"End: \"))\n",
    "        chromosome = input(\"Chromosome: (enter only the number)\")\n",
    "\n",
    "\n",
    "    GetCoordinates(hvcf, reference_fasta, start, end, chromosome)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9582f746",
   "metadata": {
    "lines_to_next_cell": 2
   },
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    try:\n",
    "        main()    \n",
    "    except KeyboardInterrupt:\n",
    "        print(\"Script interrupted by user. Exiting...\")\n",
    "        sys.exit(0)\n",
    "        raise KeyboardInterrupt"
   ]
  }
 ],
 "metadata": {
  "jupytext": {
   "cell_metadata_filter": "-all",
   "encoding": "# coding: utf-8",
   "executable": "/usr/bin/env python",
   "main_language": "python",
   "notebook_metadata_filter": "-all"
  },
  "kernelspec": {
   "display_name": "phgtools",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "name": "python",
   "version": "3.8.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
