{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "21f484b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Scrip made to detect, from a hVCF file coming from PHGv2, merging the whole pangenome, the % of \"core ranges\" and \"unique ranges\"\n",
    "# \n",
    "# functions may be increased in the future.\n",
    "# It only requires the mentioned hVCF, and indicate the number of "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "98f7ffb6",
   "metadata": {
    "lines_to_next_cell": 2
   },
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "def AmIaNotebook():\n",
    "    try:\n",
    "        shell = get_ipython().__class__.__name__\n",
    "        if shell == 'ZMQInteractiveShell':\n",
    "            return True  # Jupyter notebook or qtconsole\n",
    "        elif shell == 'TerminalInteractiveShell':\n",
    "            return False  # Terminal running IPython\n",
    "        else:\n",
    "            return False  # Other type of environment\n",
    "    except NameError:\n",
    "        return False  # Not in an interactive environment\n",
    "\n",
    "#Check if the name of this file ends with .py or with .ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a2734673",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "def NumberOfGenomes (hvcf_file):\n",
    "\n",
    "    with open(hvcf_file, 'r') as hvcf:\n",
    "        lines = hvcf.readlines()\n",
    "        num_of_genomes = 0\n",
    "        for line in lines:\n",
    "            if line.startswith('#'):\n",
    "                continue\n",
    "            else: \n",
    "                line = line.strip()\n",
    "                line = line.split('\\t')\n",
    "                keys = line[4].split(',')\n",
    "                if len(keys) > num_of_genomes:\n",
    "                    num_of_genomes = len(keys)\n",
    "                #save max value of length of keys\n",
    "\n",
    "    return num_of_genomes\n",
    "\n",
    "\n",
    "                \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a02ec005",
   "metadata": {
    "lines_to_next_cell": 2
   },
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "def DetermineCoreRanges(hvcf_file):\n",
    "\n",
    "    num_of_genomes = NumberOfGenomes(hvcf_file)\n",
    "\n",
    "    genome_numbers = list(range(1, num_of_genomes + 1))\n",
    "\n",
    "    print(genome_numbers)\n",
    "\n",
    "    with open(hvcf_file, 'r') as hvcf:\n",
    "        lines = hvcf.readlines()\n",
    "        total_ranges = 0\n",
    "        core_ranges = 0\n",
    "        unic_ranges = 0\n",
    "\n",
    "        for line in lines:\n",
    "            if line.startswith('#'):\n",
    "                continue\n",
    "            else: \n",
    "                total_ranges += 1\n",
    "                line = line.strip()\n",
    "                line = line.split('\\t')\n",
    "                keys = line[4].split(',')\n",
    "                if len(keys) == num_of_genomes:\n",
    "                    core_ranges += 1\n",
    "                if len(keys) == 1:\n",
    "                    unic_ranges += 1\n",
    "                    \n",
    "\n",
    "        print(\"\\n\")\n",
    "        print(\"here you have the whole summary:\\n \")\n",
    "\n",
    "    with open(hvcf_file, 'r') as hvcf:\n",
    "        lines = hvcf.readlines()\n",
    "        accesion_count = 0\n",
    "        dict_ranges_count = {}\n",
    "        for genome_number in genome_numbers:\n",
    "            #print(f'Number of genomes: {genome_number}')\n",
    "            count = 0\n",
    "\n",
    "            for line in lines:\n",
    "                if line.startswith('#'):\n",
    "                    continue\n",
    "                else:\n",
    "                    line = line.split('\\t')\n",
    "                    keys = line[4].split(',')\n",
    "                    #print(keys)\n",
    "                    #print(len(keys))\n",
    "                    #print(genome_number)\n",
    "                    if len(keys) == genome_number:\n",
    "                        count += 1\n",
    "\n",
    "            if genome_number is not 1 and genome_number is not num_of_genomes:\n",
    "                accesion_count += count\n",
    "\n",
    "            dict_ranges_count[genome_number] = count\n",
    "\n",
    "\n",
    "            print(f'Number of ranges with {genome_number} genomes: {count}')\n",
    "\n",
    "        print('\\nTotal number of ranges: ', total_ranges)\n",
    "        print('\\nNumber of genomes: ', num_of_genomes)\n",
    "        print('Number of core ranges: ', core_ranges)\n",
    "        print(\"Number of accesion genes are: \", accesion_count)\n",
    "        print('Number of unique ranges: ', unic_ranges)\n",
    "\n",
    "    #print(\"\\nThe sum of core, unique and accesion ranges is:\")\n",
    "    #print(core_ranges + unic_ranges + accesion_count)\n",
    "    \n",
    "    if total_ranges != core_ranges + unic_ranges + accesion_count:\n",
    "        raise ValueError('The sum of core, unique and accesion ranges does not match the total number of ranges')\n",
    "\n",
    "    return total_ranges, unic_ranges, core_ranges, accesion_count, dict_ranges_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e920ef0e",
   "metadata": {
    "lines_to_next_cell": 2
   },
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "#lets plot it\n",
    "\n",
    "def plot_pangenome_ranges_intersection(hvcf_file):\n",
    "\n",
    "    total_ranges, unic_ranges, core_ranges, accesion_count, dict_ranges_count = DetermineCoreRanges(hvcf_file)\n",
    "    num_of_genomes = NumberOfGenomes(hvcf_file)\n",
    "\n",
    "    labels = 'Core', 'Unique', 'Accesion'\n",
    "    sizes = [core_ranges, unic_ranges, accesion_count]\n",
    "    colors = ['gold', 'yellowgreen', 'lightcoral']\n",
    "    explode = (0.1, 0, 0)  # explode 1st slice\n",
    "\n",
    "    plt.pie(sizes, explode=explode, labels=labels, colors=colors, autopct='%1.1f%%', shadow=True, startangle=140)\n",
    "\n",
    "    plt.axis('equal')\n",
    "    plt.savefig(f'{hvcf_file}_1.png')\n",
    "    plt.show()\n",
    "\n",
    "    #get a subplot for accesion genes\n",
    "\n",
    "    fig, ax = plt.subplots()\n",
    "    ax.bar(dict_ranges_count.keys(), dict_ranges_count.values())\n",
    "    ax.set_xlabel('Number of genomes')\n",
    "    ax.set_ylabel('Number of ranges')\n",
    "    ax.set_title('Pangenome ranges')\n",
    "    #set color of first value to yellow, then green and last one red\n",
    "    ax.get_children()[0].set_color('yellowgreen')\n",
    "    for i in range(1, num_of_genomes):\n",
    "        ax.get_children()[i].set_color('lightcoral')\n",
    "    ax.get_children()[(num_of_genomes) -1].set_color('gold')\n",
    "    ax.set_xticks(list(dict_ranges_count.keys()))\n",
    "    ax.set_xticklabels(list(dict_ranges_count.keys()))\n",
    "\n",
    "    plt.savefig(f'{hvcf_file}_2.png')\n",
    "    plt.show()\n",
    "    print(f'The plot has been saved as a png file as {hvcf_file}_1.png and {hvcf_file}_2.png')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ce3806b",
   "metadata": {
    "lines_to_next_cell": 2
   },
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "import os\n",
    "import matplotlib.pyplot as plt\n",
    "import argparse\n",
    "import sys\n",
    "\n",
    "def main():\n",
    "    if AmIaNotebook() == False:\n",
    "        parser = argparse.ArgumentParser(description=main.__doc__)\n",
    "        parser.add_argument('--pangenome-hvcf', '-hvcf', help='Input hvcf file', required=True)\n",
    "        args = parser.parse_args()\n",
    "\n",
    "        hvcf_file = args.pangenome_hvcf\n",
    "\n",
    "    else:\n",
    "        hvcf_file = \"/scratch/PHG/output/vcf_files/merged_hvcfs_19092024.h.vcf\"\n",
    "\n",
    "    plot_pangenome_ranges_intersection(hvcf_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a7988389",
   "metadata": {
    "lines_to_next_cell": 2
   },
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "        try:\n",
    "            main()    \n",
    "        except KeyboardInterrupt:\n",
    "            print(\"Script interrupted by user. Exiting...\")\n",
    "            sys.exit(0)\n",
    "            raise KeyboardInterrupt"
   ]
  }
 ],
 "metadata": {
  "jupytext": {
   "cell_metadata_filter": "-all",
   "encoding": "# coding: utf-8",
   "executable": "/usr/bin/env python",
   "main_language": "python",
   "notebook_metadata_filter": "-all"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
