{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c461cdd",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Code to process an h.vcf imputated at the database of a Practical haplotype Graph v2 to plot the ideogram\n",
    "# \n",
    "# Performed from output of:\n",
    "# phg version 2.4.4.158\n",
    "# \n",
    "# This code has been developed at 19 sept 2024\n",
    "# Is not in the github repository\n",
    "# \n",
    "# run with pygenometracks conda, in python env\n",
    "# "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8895254b",
   "metadata": {
    "lines_to_next_cell": 2
   },
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "def AmIaNotebook():\n",
    "    try:\n",
    "        shell = get_ipython().__class__.__name__\n",
    "        if shell == 'ZMQInteractiveShell':\n",
    "            return True  # Jupyter notebook or qtconsole\n",
    "        elif shell == 'TerminalInteractiveShell':\n",
    "            return False  # Terminal running IPython\n",
    "        else:\n",
    "            return False  # Other type of environment\n",
    "    except NameError:\n",
    "        return False  # Not in an interactive environment\n",
    "\n",
    "#Check if the name of this file ends with .py or with .ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "886cce51",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "def assign_colors(genomes, reference_name):\n",
    "    \"\"\"\n",
    "    Assign a color to each genome in the pangenome, up to 20 genomes at this moment. Reference is also assigned as Black\n",
    "    You need to provide the list of genomes and the reference genome name    \n",
    "    \"\"\"\n",
    "    colors = {\n",
    "    \"red\": \"255,0,0\",\n",
    "    \"green\": \"0,255,0\",\n",
    "    \"blue\": \"0,0,255\",\n",
    "    \"yellow\": \"255,255,0\",\n",
    "    \"cyan\": \"0,255,255\",\n",
    "    \"magenta\": \"255,0,255\",\n",
    "    \"brown\": \"165,42,42\",\n",
    "    \"gray\": \"128,128,128\",\n",
    "    \"orange\": \"255,165,0\",\n",
    "    \"purple\": \"128,0,128\",\n",
    "    \"pink\": \"255,192,203\",\n",
    "    \"lime\": \"0,255,0\",\n",
    "    \"navy\": \"0,0,128\",\n",
    "    \"teal\": \"0,128,128\",\n",
    "    \"olive\": \"128,128,0\",\n",
    "    \"maroon\": \"128,0,0\",\n",
    "    \"violet\": \"238,130,238\",\n",
    "    \"gold\": \"255,215,0\",\n",
    "    \"silver\": \"192,192,192\",\n",
    "    \"beige\": \"245,245,220\",\n",
    "    \"coral\": \"255,127,80\",\n",
    "    \"indigo\": \"75,0,130\",\n",
    "    \"khaki\": \"240,230,140\",\n",
    "    \"lavender\": \"230,230,250\",\n",
    "    }\n",
    "\n",
    "\n",
    "    colors = dict(list(colors.items())[:len(genomes)])\n",
    "    colors_dic = {genomes[i]: color for i, color in enumerate(colors.values())}\n",
    "\n",
    "    if reference_name not in genomes:\n",
    "        ref_color = {f\"{reference_name}\": \"0,0,0\"}\n",
    "\n",
    "    print(f\"Reference genome color: {ref_color}, saved as ref_color\")\n",
    "    print(f\"Genomes colors: {colors_dic}, saved as colors_dic\")\n",
    "\n",
    "    return colors_dic, ref_color\n",
    "\n",
    "\n",
    "#colors_dict, ref_color = assign_colors(genomes, reference_name)\n",
    "\n",
    "#print (f\"Colors assigned to genomes: {colors_dict}\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "861e686e",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "def hvcf2bed(hvcf, plot_folder, colors_dic):\n",
    "    \"\"\"\n",
    "    This function will convert the hvcf file to a bed file, and store it at the plot folder. BED file will have the same name as hvcf. \n",
    "    It will use the coordinates of the reference.\n",
    "    \n",
    "    The columns are:\n",
    "    1. NameOfHVCF       #Doing this for ploting software, is not strictly accurate with BED format\n",
    "    2. Start\n",
    "    3. End\n",
    "    4. Chr             #Doing this for ploting and filtrering in this software\n",
    "    5. Score (actually a dummy score)\n",
    "    6. Strand\n",
    "    7. thickStart\n",
    "    8. thickEnd\n",
    "    9. itemRgb\n",
    "\n",
    "    *Notice that the new files WILL NOT CONTAIN in each line to which genome is the range belonging to.\n",
    "    So in further steps, you must convert from the color dictionary the RGB color to each genome of the pangenome :)\n",
    "    \"\"\"\n",
    "\n",
    "    if not os.path.exists(plot_folder):\n",
    "        os.makedirs(plot_folder)\n",
    "        print(f\"Created directory: {plot_folder}\")\n",
    "\n",
    "    \n",
    "    hvcf_BED = (f\"{plot_folder}TEMP_{os.path.basename(hvcf).split('.')[0]}.bed\")\n",
    "\n",
    "    if os.path.exists(hvcf_BED):\n",
    "        #print(f\"File {hvcf_BED} already exists, removing it and building it again\") \n",
    "        os.remove(hvcf_BED)\n",
    "\n",
    "    print(f\"Converting {hvcf} to BED file\")\n",
    "\n",
    "    if hvcf.endswith('.gz'):                        ##Checks if the file is compressed or not and acts depending on it\n",
    "        open_func = lambda f: gzip.open(f, 'rt')\n",
    "    else:\n",
    "        open_func = open\n",
    "\n",
    "    with open_func(hvcf) as hvcf_file, open(hvcf_BED, 'w') as bed_file:\n",
    "        first_line = True\n",
    "        lines = hvcf_file.readlines()\n",
    "        for line in lines:\n",
    "            \n",
    "            if line.startswith ('##ALT=<ID'):\n",
    "                line = line.strip().split(',')\n",
    "                range = line[-1]                   #take last element. Sometimes there is more than one range at the haplotype itself, so we take the reference one\n",
    "                                                #Actually this bug should be fixed at the PHG version 2.4.8.162, but lets keep this code for older hvcf files\n",
    "                range = range.split('=')[1]\n",
    "                chr = range.split(':')[0]\n",
    "                chr = chr.split (\"_\")[0]\n",
    "                start = range.split(':')[1].split('-')[0]\n",
    "                end = range.split(':')[1].split('-')[1]\n",
    "                end = end.split('>')[0]\n",
    "                name = line[1].split(':')[1]\n",
    "                name = name.split('\"')[0].replace(\" \", \"\")\n",
    "                #print(name, chr, start, end, name)\n",
    "\n",
    "                if first_line == True:\n",
    "                    bed_file.write(\"chrom\\tchromStart\\tchromEnd\\tname\\tscore\\tstrand\\tthickStart\\tthickEnd\\titemRgb\\n\")\n",
    "                    first_line = False\n",
    "                    bed_file.write(f\"{os.path.basename(hvcf).split('.')[0]}\\t{start}\\t{end}\\t{chr}\\t100\\t+\\t{start}\\t{end}\\t{colors_dic[name]}\\n\")\n",
    "                    #print(line)\n",
    "                else:\n",
    "                    bed_file.write(f\"{os.path.basename(hvcf).split('.')[0]}\\t{start}\\t{end}\\t{chr}\\t100\\t+\\t{start}\\t{end}\\t{colors_dic[name]}\\n\")\n",
    "                    #print(chr)\n",
    "                    pass\n",
    "            \n",
    "\n",
    "        print(f\"finished processing {hvcf} to BED file {hvcf_BED}\")\n",
    "      \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5c872cd3",
   "metadata": {
    "lines_to_next_cell": 2
   },
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "def CaptureChrFromBED (hvcf, plot_folder):\n",
    "    \"\"\"\n",
    "    This function will capture the chromosome from the BED file, storing them in a list\n",
    "    \"\"\"\n",
    "    hvcf_BED = (f\"{plot_folder}TEMP_{os.path.basename(hvcf).split('.')[0]}.bed\")\n",
    "    chr_list = []\n",
    "    with open (hvcf_BED, 'r') as f:\n",
    "        lines = f.readlines()\n",
    "        lines = lines[1:]\n",
    "        for line in lines:\n",
    "            chr = line.split('\\t')[3]\n",
    "            if chr not in chr_list:\n",
    "                chr_list.append(chr)\n",
    "    chr_list.sort()\n",
    "    print(f\"Chromosomes found in the BED file are: {chr_list}\")\n",
    "    return chr_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8b45c4b6",
   "metadata": {
    "lines_to_next_cell": 2
   },
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "def BEDSplitByChr (hvcf, plot_folder):\n",
    "    pass\n",
    "    \"\"\"\n",
    "    splits and sorts the BED file in different files by chromosome, sorting it by start position. \n",
    "    This is only in order to plot it with pygenometracks. Takes by output the BED file generated by hvcf2bed.\n",
    "\n",
    "    \"\"\"\n",
    "    hvcf_BED = (f\"{plot_folder}TEMP_{os.path.basename(hvcf).split('.')[0]}.bed\")\n",
    "\n",
    "    chr_list = CaptureChrFromBED(hvcf, plot_folder)\n",
    "    print(\"Command CaptureChrFromBED executed\")\n",
    "\n",
    "    for chr in chr_list:    #check if the files splited by chr already exist, and if they do, remove them            \n",
    "        chr_file = (f\"{plot_folder}TEMP_{os.path.basename(hvcf).split('.')[0]}_{chr}.bed\")\n",
    "        if os.path.exists(chr_file):\n",
    "            print(f\"File {chr_file} already exists, removing it and building it again\") \n",
    "            os.remove (chr_file)\n",
    "\n",
    "    try:\n",
    "        open (hvcf_BED, 'r')\n",
    "    except FileNotFoundError:\n",
    "        print(f\"File {hvcf_BED} not found\")\n",
    " \n",
    "    with open (hvcf_BED, 'r') as f:\n",
    "        lines = f.readlines()\n",
    "        if len(lines) == 0:\n",
    "            raise Exception(f\"File {hvcf_BED} is empty\")\n",
    "        if len(chr_list) == 0:\n",
    "            raise Exception(f\"Chromosome list is empty\")\n",
    "        \n",
    "        print(f\"opening {hvcf_BED} to split it by chromosome\")\n",
    "        \n",
    "        for chr in chr_list:\n",
    "            chr_file = (f\"{plot_folder}TEMP_{os.path.basename(hvcf).split('.')[0]}_{chr}.bed\")\n",
    "            with open (chr_file, 'w') as chr_bed:\n",
    "                chr_bed.write(\"chrom\\tchromStart\\tchromEnd\\tname\\tscore\\tstrand\\tthickStart\\tthickEnd\\titemRgb\\n\")\n",
    "                for line in lines:\n",
    "                    if chr in line:\n",
    "                        chr_bed.write(line)\n",
    "            try:\n",
    "                open(chr_file, 'r')\n",
    "            except FileNotFoundError:\n",
    "                print(f\"File {chr_file} not found\") \n",
    "            else:\n",
    "                if len(open(chr_file).readlines()) <= 1:  \n",
    "                    raise Exception(f\"File {chr_file} of the {chr} is empty\\n Im actually trying to find the chromosome {chr} in the file {hvcf_BED}\")\n",
    "                print(f\"Finished splitting {hvcf_BED} by chromosome to {chr}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "282a3f19",
   "metadata": {
    "lines_to_next_cell": 2
   },
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "def sortBED (hvcf, plot_folder):\n",
    "\n",
    "    \"\"\"\n",
    "Function to sort the files, inside hvcf plotting.\n",
    "Requires the hvcf file, ploting folder, and calls the function CaptureChrFromBed\n",
    "    \"\"\"\n",
    "\n",
    "    chr_list = CaptureChrFromBED(hvcf, plot_folder)\n",
    "\n",
    "\n",
    "#Now sort the files by start position\n",
    "    for chr in chr_list:\n",
    "        chr_file = (f\"{plot_folder}TEMP_{os.path.basename(hvcf).split('.')[0]}_{chr}.bed\")\n",
    "        sorted_chr_file = (f\"{plot_folder}TEMP_{os.path.basename(hvcf).split('.')[0]}_{chr}_sorted.bed\")\n",
    "        if os.path.exists(sorted_chr_file):\n",
    "            print(f\"File {sorted_chr_file} already exists, removing it and building it again\") \n",
    "            os.remove(sorted_chr_file)\n",
    "\n",
    "        try:    \n",
    "            open (chr_file, 'r')\n",
    "        except FileNotFoundError:\n",
    "            print(f\"File {chr_file} not found\") \n",
    "        else:\n",
    "            lines = len(open(chr_file).readlines())\n",
    "            if lines == 0:\n",
    "                raise Exception(f\"File {chr_file} is empty\")\n",
    "\n",
    "        with open (chr_file, 'r') as f:\n",
    "            lines = f.readlines()\n",
    "            lines = lines[1:]\n",
    "            lines.sort(key=lambda x: int(x.split('\\t')[1]))\n",
    "            with open (sorted_chr_file, 'w') as sorted_chr_bed:\n",
    "                sorted_chr_bed.write(\"chrom\\tchromStart\\tchromEnd\\tname\\tscore\\tstrand\\tthickStart\\tthickEnd\\titemRgb\\n\")\n",
    "                for line in lines:\n",
    "                    sorted_chr_bed.write(line)\n",
    "\n",
    "                    try:\n",
    "                        line = line.split('\\t')\n",
    "                        len(line) == 9  \n",
    "                    except:\n",
    "                        raise Exception(f\"Line {line} does not have 9 columns\")\n",
    "\n",
    "            try:\n",
    "                with open(sorted_chr_file, 'r') as f:\n",
    "                    lines = f.readlines\n",
    "            except FileNotFoundError:\n",
    "                print(f\"File {sorted_chr_file} not found\")\n",
    "            else:\n",
    "                if len(open(sorted_chr_file).readlines()) <= 1:\n",
    "                    raise Exception(f\"File {sorted_chr_file} is empty\")\n",
    "\n",
    "\n",
    "        try:\n",
    "            with open(sorted_chr_file, 'r')  as f:\n",
    "                lines = f.readlines\n",
    "        except FileNotFoundError:\n",
    "            print(f\"File {sorted_chr_file} not found\")   \n",
    "        else:\n",
    "            lines = len(open(sorted_chr_file).readlines())\n",
    "            if lines == 0:\n",
    "                raise Exception(f\"File {sorted_chr_file} is empty\")\n",
    "        #print(f\"Finished sorting {chr_file} to {sorted_chr_file}, and removing the first one\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01b7ad83",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "def PrepareTrack_Hvcf2plot(hvcf, reference_name, plot_folder):\n",
    "    \"\"\"\n",
    "    This function will prepare the tracks for pygenometracks, so it can plot the hvcf files. Indicate the hvcf initial file, and the reference name, as long\n",
    "    as the folder to save the plot.\n",
    "    However, it actually uses the bed files from BEDSplitByChr function.\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    chr_list = CaptureChrFromBED(hvcf, plot_folder)\n",
    "    print(chr_list)\n",
    "\n",
    "    #Now we prepare the tracks for pygenometracks\n",
    "\n",
    "    track_ini = (f\"{plot_folder}TEMP_{os.path.basename(hvcf).split('.')[0]}_track.ini\")\n",
    "\n",
    "    if os.path.exists(track_ini):\n",
    "        print(f\"File {track_ini} already exists, removing it and building it again\") \n",
    "        os.remove(track_ini)\n",
    "\n",
    "    with open (track_ini, 'w') as track:\n",
    "        \n",
    "        for chr in chr_list:\n",
    "            track.write(f\"[haplotype_{chr}]\\n\")\n",
    "            track.write(f\"title = {chr}\\n\")\n",
    "            track.write(f\"file = {plot_folder}TEMP_{os.path.basename(hvcf).split('.')[0]}_{chr}_sorted.bed \\n\")\n",
    "            track.write(f\"height = 3\\n\")\n",
    "            track.write(f\"color = bed_rgb\\n\")\n",
    "            track.write(\"display = collapsed\\n\")\n",
    "            track.write(\"labels = false\\n\")\n",
    "            track.write(\"label_fontsize = 10\\n\")\n",
    "            track.write(\"border_color = black\\n\")\n",
    "            track.write(\"line_width = 0.05\\n\")\n",
    "            track.write(\"\\n\")\n",
    "            track.write(\"[spacer]\\nheight = 0.5\\n\")\n",
    "\n",
    "        track.write(\"[x-axis]\\n\")\n",
    "        track.write(\"where = bottom\\n\")\n",
    "        #track.write(\"label = true\\n\")\n",
    "        #track.write(\"font_size = 10\\n\")\n",
    "        #track.write(f\"title = Mediterranean pangenome {chromosome_to_plot} using as reference MorexV3\\n\")\n",
    "        track.write(\"[spacer]\\nheight = 0.5\\n\")\n",
    "\n",
    "    print(f\"Finished preparing the tracks for pygenometracks at {track_ini}\")\n",
    "    return track_ini\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "26e68686",
   "metadata": {
    "lines_to_next_cell": 2
   },
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "def SetExtremesToPlot(hvcf, plot_folder):\n",
    "    \"\"\"\n",
    "    \"\"\"    \n",
    "    #capturing the extreme values to plot of each chromosome\n",
    "    chr_list = CaptureChrFromBED(hvcf, plot_folder)\n",
    "    print(chr_list)\n",
    "    chr_max = None\n",
    "    for chr in chr_list:\n",
    "        chr_file = (f\"{plot_folder}TEMP_{os.path.basename(hvcf).split('.')[0]}_{chr}_sorted.bed\")\n",
    "        with open (chr_file, 'r') as f:\n",
    "            #print (f\"Processing {chr_file}\")\n",
    "            lines = f.readlines()\n",
    "            if lines == []:\n",
    "                raise Exception(f\"Empty file {chr_file}\")\n",
    "                \n",
    "            try:    \n",
    "                last_line = lines[-1]\n",
    "            except:\n",
    "                print(f\"This is the file im working now: {lines}\\n\")\n",
    "                raise IndexError(f\"I can not find the last line in {chr_file}\")\n",
    "\n",
    "            last_line = lines[-1]\n",
    "            last_coord = last_line.split('\\t')[2]\n",
    "            if chr_max == None or chr_max < last_coord:\n",
    "                chr_max = last_coord\n",
    "    chr_max = int(int(chr_max)*1.005)\n",
    "    print(f\"Max coord to plot will be {chr_max}\")\n",
    "\n",
    "    return chr_max"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cad09ee5",
   "metadata": {
    "lines_to_next_cell": 2
   },
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "def RunPygenometracksCommandHVCF(hvcf, reference_name, plot_folder, genomes):\n",
    "\n",
    "    \"\"\"\n",
    "    Building and executing the command for plotting:\n",
    "    It calls the functions SetExtremesToPlot and PrepareTrack_HvcfPlot\n",
    "    \"\"\"\n",
    "\n",
    "    chr_max = SetExtremesToPlot(hvcf, plot_folder)\n",
    "    track_ini = PrepareTrack_Hvcf2plot(hvcf, reference_name, plot_folder)\n",
    "    chr_list = CaptureChrFromBED(hvcf, plot_folder)\n",
    "\n",
    "    width = len(genomes)*12\n",
    "    height = len(chr_list)*10\n",
    "\n",
    "    command = (f\"pyGenomeTracks --tracks {track_ini} --region {os.path.basename(hvcf).split('.')[0]}:1-{chr_max} --outFileName {plot_folder}/{os.path.basename(hvcf)}.png --width {width} --dpi 500\")\n",
    "\n",
    "    try: \n",
    "        result = subprocess.run(command, shell=True, check=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True)\n",
    "        print(f\"running the command {command}\")\n",
    "        #print(result.stdout)\n",
    "        #print(result.stderr)\n",
    "        if \"No valid intervals were found\" in result.stdout or \"No valid intervals were found\" in result.stderr:\n",
    "            raise Exception(f\"Command {command} failed with error: {result.stderr}\")\n",
    "    except subprocess.CalledProcessError as e:\n",
    "        raise Exception(f\"Command {command} failed with error: {e.stderr}\")\n",
    "        \n",
    "    print(f\"the plot is saved at {plot_folder}{os.path.basename(hvcf).split('.')[0]}.png\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ab784e0",
   "metadata": {
    "lines_to_next_cell": 2
   },
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "def AddLegendToPygenometracks (hvcf, plot_folder, genomes, reference_name):\n",
    "    \"\"\"\n",
    "    This function will add a legend to the pygenometracks plot, with the colors assigned to each genome.\n",
    "    \"\"\"\n",
    "\n",
    "    print (\"lets add a legend to the plot\")\n",
    "\n",
    "    # Transform the colors RGB\n",
    "    colors_dict, ref_color = assign_colors(genomes, reference_name)\n",
    "    num_genomes = len(genomes)\n",
    "    # Convert colors from RGB to hex\n",
    "    for genome in colors_dict:\n",
    "        r, g, b = map(int, colors_dict[genome].split(','))\n",
    "        colors_dict[genome] = f'#{r:02X}{g:02X}{b:02X}'\n",
    "    r, g, b = map(int, ref_color[reference_name].split(','))\n",
    "    ref_color[reference_name] = f'#{r:02X}{g:02X}{b:02X}'\n",
    "\n",
    "    # Create legend elements\n",
    "    legend_elements = [Line2D([0], [0], marker='o', color='w', label=genome, \n",
    "                            markerfacecolor=colors_dict[genome], markersize=num_genomes*20)  \n",
    "                    for genome in colors_dict]\n",
    "    legend_elements.append(Line2D([0], [0], marker='o', color='w', label=reference_name, \n",
    "                                markerfacecolor=ref_color[reference_name], markersize=num_genomes*20))\n",
    "\n",
    "    # Create a new figure for the legend\n",
    "    fig, ax = plt.subplots(figsize=(10, 10))  # Adjust size as necessary\n",
    "    ax.legend(handles=legend_elements, loc='center', frameon=False,fontsize=num_genomes*20)\n",
    "\n",
    "    # Remove axes for a clean legend\n",
    "    ax.axis('off')\n",
    "\n",
    "    # Save the legend as an image\n",
    "    plt.savefig(f'{plot_folder}TEMP_legend.png', bbox_inches='tight', transparent=True)\n",
    "    plt.close()\n",
    "\n",
    "    # Open the main plot and the legend, but only if you are in a notebook\n",
    "    plot_image = Image.open(f'{plot_folder}{os.path.basename(hvcf)}.png')\n",
    "    legend_image = Image.open(f'{plot_folder}TEMP_legend.png')\n",
    "    if AmIaNotebook():\n",
    "        display(plot_image)\n",
    "        display(legend_image)\n",
    "\n",
    "    # Get the dimensions of the plot and legend\n",
    "    plot_width, plot_height = plot_image.size\n",
    "    legend_width, legend_height = legend_image.size\n",
    "\n",
    "    # Define the position where you want to place the legend (e.g., bottom-right corner)\n",
    "    position = (plot_width - legend_width - 10, plot_height - legend_height - 10)  # 10-pixel padding\n",
    "\n",
    "    # Paste the legend onto the plot\n",
    "    plot_image.paste(legend_image, position, legend_image)\n",
    "\n",
    "    # Save the final merged image\n",
    "    plot_image.save(f'{plot_folder}{os.path.basename(hvcf)}.png')\n",
    "\n",
    "    print (\"Done! The legend is added to the plot\\n\")\n",
    "    print(f\"Your plot is saved at {plot_folder}{os.path.basename(hvcf)}.png\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23397605",
   "metadata": {
    "lines_to_next_cell": 2
   },
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "import os\n",
    "import gzip\n",
    "import re\n",
    "import subprocess\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.patches import Patch\n",
    "from PIL import Image\n",
    "from matplotlib.lines import Line2D\n",
    "import argparse\n",
    "import sys\n",
    "\n",
    "def main():\n",
    "    \"\"\"\n",
    "    This function will plot the hvcf file using pygenometracks. It will use the track_ini file generated by PrepareTrack_Hvcf2plot\n",
    "    \"\"\"\n",
    "    if AmIaNotebook() == False:\n",
    "        parser = argparse.ArgumentParser(description=main.__doc__)\n",
    "        parser.add_argument(\"--input-hvcf\", \"-hvcf\", help=\"Path to the hvcf file\")\n",
    "        parser.add_argument(\"--pangenome-hvcf-folder\", \"-folder\", help=\"Path to the folder where the plots will be saved\")\n",
    "        parser.add_argument(\"--reference-hvcf\", \"-ref\", help=\"Path to the reference hvcf file\")\n",
    "        args = parser.parse_args()\n",
    "\n",
    "        hvcf = args.input_hvcf\n",
    "        pangenome_hvcf = args.pangenome_hvcf_folder\n",
    "        reference_hvcf = args.reference_hvcf\n",
    "\n",
    "    else:\n",
    "        pangenome_hvcf = \"/scratch/PHG/output/vcf_files\"\n",
    "        hvcf = \"/scratch/PHG/output/ensambled_genomes/1740D-268-01_S1_L001_R1_001_GDB136.h.vcf\"\n",
    "        reference_hvcf = \"/scratch/PHG/vcf_dbs/hvcf_files/MorexV3.h.vcf.gz\"\n",
    "\n",
    "\n",
    "\n",
    "    plot_folder = (f\"{pangenome_hvcf}/plots/\")\n",
    "    hvcf_files_path = [os.path.join(pangenome_hvcf, file) for file in os.listdir(pangenome_hvcf) if file.endswith('.h.vcf.gz')]\n",
    "    genomes = [os.path.basename(file).split('.')[0] for file in hvcf_files_path]\n",
    "    num_genomes = len(genomes)\n",
    "    hvcf_files = [os.path.basename(file) for file in hvcf_files_path]\n",
    "    print(f\"Your hvcf file is: \\n{hvcf}\\n The genomes at pangenome are currently {num_genomes}: \\n{genomes}\\n \")\n",
    "    reference_name = os.path.basename(reference_hvcf).split('.')[0]\n",
    "    print(f\"Your reference hvcf file is: \\n{reference_hvcf}\\n The reference genome is: \\n{reference_name}\\n \")\n",
    "\n",
    "    print(f\"the plot folder is {plot_folder}\")  \n",
    "\n",
    "    colors_dic, ref_color = assign_colors(genomes, reference_name)\n",
    "\n",
    "    hvcf2bed(hvcf, plot_folder, colors_dic)\n",
    "\n",
    "    BEDSplitByChr (hvcf, plot_folder)\n",
    "\n",
    "    sortBED (hvcf, plot_folder)\n",
    "\n",
    "    RunPygenometracksCommandHVCF(hvcf, reference_name, plot_folder, genomes)\n",
    "\n",
    "    AddLegendToPygenometracks (hvcf, plot_folder, genomes, reference_name)\n",
    "\n",
    "    #Remove temporary files\n",
    "    chr_list = CaptureChrFromBED(hvcf, plot_folder)\n",
    "    os.remove(f\"{plot_folder}TEMP_{os.path.basename(hvcf).split('.')[0]}.bed\")\n",
    "    os.remove(f\"{plot_folder}TEMP_{os.path.basename(hvcf).split('.')[0]}_track.ini\")\n",
    "    os.remove(f\"{plot_folder}TEMP_legend.png\")\n",
    "    for chr in chr_list:\n",
    "        os.remove(f\"{plot_folder}TEMP_{os.path.basename(hvcf).split('.')[0]}_{chr}.bed\")\n",
    "        os.remove(f\"{plot_folder}TEMP_{os.path.basename(hvcf).split('.')[0]}_{chr}_sorted.bed\")\n",
    "\n",
    "    print(\"Temporary files removed\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e4682a9a",
   "metadata": {
    "lines_to_next_cell": 2
   },
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "        try:\n",
    "            main()    \n",
    "        except KeyboardInterrupt:\n",
    "            print(\"Script interrupted by user. Exiting...\")\n",
    "            sys.exit(0)\n",
    "            raise KeyboardInterrupt"
   ]
  }
 ],
 "metadata": {
  "jupytext": {
   "cell_metadata_filter": "-all",
   "encoding": "# coding: utf-8",
   "executable": "/usr/bin/env python",
   "main_language": "python",
   "notebook_metadata_filter": "-all"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
