{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "69e526d6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d6b3a9e",
   "metadata": {
    "lines_to_next_cell": 2
   },
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "def AmIaNotebook():\n",
    "    try:\n",
    "        shell = get_ipython().__class__.__name__\n",
    "        if shell == 'ZMQInteractiveShell':\n",
    "            return True  # Jupyter notebook or qtconsole\n",
    "        elif shell == 'TerminalInteractiveShell':\n",
    "            return False  # Terminal running IPython\n",
    "        else:\n",
    "            return False  # Other type of environment\n",
    "    except NameError:\n",
    "        return False  # Not in an interactive environment\n",
    "\n",
    "#Check if the name of this file ends with .py or with .ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f38a3b0",
   "metadata": {
    "lines_to_next_cell": 2
   },
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "def DefineHVCFs(hvcf_folder, reference_hvcf):\n",
    "    \"\"\"\n",
    "    Define the haplotype VCF files to be plotted\n",
    "    \"\"\"\n",
    "    haplotype_files = [f for f in os.listdir(hvcf_folder) if f.endswith('.h.vcf.gz')]\n",
    "    if reference_hvcf in haplotype_files:\n",
    "        haplotype_files.remove(reference_hvcf)\n",
    "\n",
    "    haplotype_names = [f.split('/')[-1].split('.h.vcf.gz')[0] for f in haplotype_files]\n",
    "    reference_name = reference_hvcf.split('.h.vcf.gz')[0].split('/')[-1]\n",
    "    \n",
    "    return haplotype_files, haplotype_names, reference_name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "628cc24a",
   "metadata": {
    "lines_to_next_cell": 2
   },
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "def PangenomeColors(reference_name, haplotype_names):\n",
    "    \"\"\"\n",
    "    Define the colors for the pangenome plot\n",
    "    \"\"\"\n",
    "\n",
    "    columns_header = (\"chrom\", \"chromStart\", \"chromEnd\", \"name\", \"score\", \"strand\", \"thickStart\", \"thickEnd\", \"itemRgb\")\n",
    "\n",
    "    colors = {\n",
    "        \"red\": \"255, 0, 0\",\n",
    "        \"green\": \"0, 255, 0\",\n",
    "        \"blue\": \"0, 0, 255\",\n",
    "        \"yellow\": \"255, 255, 0\",\n",
    "        \"cyan\": \"0, 255, 255\",\n",
    "        \"magenta\": \"255, 0, 255\",\n",
    "        \"brown\": \"165, 42, 42\",\n",
    "        \"gray\": \"128, 128, 128\",\n",
    "        \"orange\": \"255, 165, 0\",\n",
    "        \"purple\": \"128, 0, 128\",\n",
    "        \"pink\": \"255, 192, 203\",\n",
    "        \"lime\": \"0, 255, 0\",\n",
    "        \"navy\": \"0, 0, 128\",\n",
    "        \"teal\": \"0, 128, 128\",\n",
    "        \"olive\": \"128, 128, 0\",\n",
    "        \"maroon\": \"128, 0, 0\",\n",
    "        \"violet\": \"238, 130, 238\",\n",
    "        \"gold\": \"255, 215, 0\",\n",
    "        \"silver\": \"192, 192, 192\",\n",
    "        \"beige\": \"245, 245, 220\",\n",
    "        \"coral\": \"255, 127, 80\",\n",
    "        \"indigo\": \"75, 0, 130\",\n",
    "        \"khaki\": \"240, 230, 140\",\n",
    "        \"lavender\": \"230, 230, 250\",\n",
    "    }\n",
    "\n",
    "    ref_color = {f\"{reference_name}\": \"0, 0, 0\"}\n",
    "\n",
    "    #Only use the first colors, equal to number of haplotype_names\n",
    "    colors = dict(list(colors.items())[:len(haplotype_names)])\n",
    "\n",
    "    # Create a new dictionary with identifiers as keys\n",
    "    colors_dic = {haplotype_names[i]: color for i, color in enumerate(colors.values())}\n",
    "\n",
    "    print(f\"\\nThe colors assigned to the haplotypes are: {colors_dic}\\n\")\n",
    "\n",
    "    return colors_dic, ref_color, columns_header"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "982092c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "def CreateBedFile(hvcf_folder, plots_folder, haplotype_names, reference_name, reference_hvcf, colors_dic, columns_header):\n",
    "    \"\"\"\n",
    "    Create the bed files for each genome haplotype\n",
    "    \"\"\"\n",
    "\n",
    "    for file in haplotype_names:\n",
    "        #check if file already exists:\n",
    "        print(f\"starting to process {file}...\")\n",
    "        if os.path.exists(f\"{hvcf_folder}plots/{file}_haplotype.bed\"):\n",
    "            print(f\"file {file}_haplotype.bed already exists, removing and making it again...\")\n",
    "            os.remove(f\"{hvcf_folder}plots/{file}_haplotype.bed\")\n",
    "            first_line = True\n",
    "            with gzip.open(f\"{hvcf_folder}{file}.h.vcf.gz\", 'rt', encoding='utf-8', errors='ignore') as vcf_file, open(f\"{hvcf_folder}plots/{file}_haplotype.bed\", 'w') as output:\n",
    "                lines = vcf_file.readlines()\n",
    "                contig_haplotypes = [line for line in lines if not line.startswith(\"#\")]\n",
    "                for haplotype_line in contig_haplotypes:\n",
    "                    haplotype_line = haplotype_line.strip().split('\\t')\n",
    "                    chr = haplotype_line[0]\n",
    "                    start = haplotype_line[1]\n",
    "                    end = haplotype_line[7].split('=')[1]\n",
    "                    key = haplotype_line[4]\n",
    "                    extract_key = re.search(r'<(.*?)>', key)\n",
    "                    if extract_key:\n",
    "                        key = extract_key.group(1)\n",
    "\n",
    "                    if first_line == True:\n",
    "                        output.write(\"\\t\".join(columns_header) + \"\\n\")\n",
    "                        first_line = False\n",
    "                        output.write(f\"{chr}\\t{start}\\t{end}\\t{key}\\t100\\t+\\t{start}\\t{end}\\t{colors_dic[file]}\\n\")\n",
    "                        first_line = False\n",
    "                    else:\n",
    "                        output.write(f\"{chr}\\t{start}\\t{end}\\t{key}\\t100\\t+\\t{start}\\t{end}\\t{colors_dic[file]}\\n\")\n",
    "                print(f\"finished processing {file}_haplotype.bed\")\n",
    "\n",
    "\n",
    "        else:\n",
    "            os.makedirs(plots_folder, exist_ok=True)\n",
    "            print(f\"generating {file}_haplotype.bed...\")\n",
    "            first_line = True\n",
    "            with gzip.open(f\"{hvcf_folder}{file}.h.vcf.gz\", 'rt', encoding='utf-8', errors='ignore') as vcf_file, open(f\"{hvcf_folder}plots/{file}_haplotype.bed\", 'w') as output:\n",
    "                lines = vcf_file.readlines()\n",
    "                contig_haplotypes = [line for line in lines if not line.startswith(\"#\")]\n",
    "                for haplotype_line in contig_haplotypes:\n",
    "                    haplotype_line = haplotype_line.strip().split('\\t')\n",
    "                    chr = haplotype_line[0]\n",
    "                    start = haplotype_line[1]\n",
    "                    end = haplotype_line[7].split('=')[1]\n",
    "                    key = haplotype_line[4]\n",
    "                    extract_key = re.search(r'<(.*?)>', key)\n",
    "                    if extract_key:\n",
    "                        key = extract_key.group(1)\n",
    "\n",
    "                    if first_line == True:\n",
    "                        output.write(\"\\t\".join(columns_header) + \"\\n\")\n",
    "                        first_line = False\n",
    "                        output.write(f\"{chr}\\t{start}\\t{end}\\t{key}\\t100\\t+\\t{start}\\t{end}\\t{colors_dic[file]}\\n\")\n",
    "                        first_line = False\n",
    "                    else:\n",
    "                        output.write(f\"{chr}\\t{start}\\t{end}\\t{key}\\t100\\t+\\t{start}\\t{end}\\t{colors_dic[file]}\\n\")\n",
    "                print(f\"finished processing {file}_haplotype.bed\")\n",
    "\n",
    "    #Same for reference\n",
    "\n",
    "    print(f\"starting to process {reference_name}...\")\n",
    "    if os.path.exists(f\"{hvcf_folder}plots/{reference_name}_haplotype.bed\"):\n",
    "                print(f\"file {reference_name}_haplotype.bed already exists, removing and making it again...\")\n",
    "                os.remove(f\"{hvcf_folder}plots/{reference_name}_haplotype.bed\")\n",
    "                first_line = True\n",
    "    if reference_name != None:            \n",
    "        with gzip.open(f\"{reference_hvcf}\", 'rt', encoding='utf-8', errors='ignore') as vcf_file, open(f\"{plots_folder}{reference_name}_haplotype.bed\", 'w') as output:\n",
    "            lines = vcf_file.readlines()\n",
    "            contig_haplotypes = [line for line in lines if not line.startswith(\"#\")]\n",
    "            for haplotype_line in contig_haplotypes:\n",
    "                haplotype_line = haplotype_line.strip().split('\\t')\n",
    "                chr = haplotype_line[0]\n",
    "                start = haplotype_line[1]\n",
    "                end = haplotype_line[7].split('=')[1]\n",
    "                key = haplotype_line[4]\n",
    "                extract_key = re.search(r'<(.*?)>', key)\n",
    "                if extract_key:\n",
    "                    key = extract_key.group(1)\n",
    "\n",
    "                if first_line == True:\n",
    "                    output.write(\"\\t\".join(columns_header) + \"\\n\")\n",
    "                    first_line = False\n",
    "                    output.write(f\"{chr}\\t{start}\\t{end}\\t{key}\\t100\\t+\\t{start}\\t{end}\\t{colors_dic[file]}\\n\")\n",
    "                    first_line = False\n",
    "                else:\n",
    "                    output.write(f\"{chr}\\t{start}\\t{end}\\t{key}\\t100\\t+\\t{start}\\t{end}\\t{colors_dic[file]}\\n\")\n",
    "            print(f\"finished processing {file}_haplotype.bed\")\n",
    "\n",
    "\n",
    "\n",
    "                \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a66e1e5f",
   "metadata": {
    "lines_to_next_cell": 2
   },
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "def WholeChrLenght(ref_fasta, chromosome_to_plot, reference_name, region_to_plot):\n",
    "\n",
    "    \"\"\"\n",
    "    Get the whole chromosome length only if plotting the whole chromosome\n",
    "    \"\"\"\n",
    "\n",
    "    #extract the list of real chromosomes\n",
    "    with open(ref_fasta, 'r') as fasta_file:\n",
    "        fasta_lines = fasta_file.readlines()\n",
    "        chromosomes = [line for line in fasta_lines if line.startswith(\">chr\")]\n",
    "        real_chromosomes = [chromosome.split(' ')[0].replace(\">\", \"\") for chromosome in chromosomes]\n",
    "        print(real_chromosomes)\n",
    "\n",
    "        chromosome_dict = {f\"chr{i+1}\": real_chromosomes[i] for i in range(len(real_chromosomes))}\n",
    "\n",
    "    # Print the dictionary\n",
    "    print(chromosome_dict)\n",
    "\n",
    "    if region_to_plot == None:\n",
    "\n",
    "\n",
    "        chromosome_found = False\n",
    "        chromosome_length = 0\n",
    "        print(f\"The chromosome to plot is the {chromosome_to_plot}, which actually is called {chromosome_dict[chromosome_to_plot]} in {reference_name}...\")\n",
    "\n",
    "        for line in fasta_lines:\n",
    "            if not line.startswith (f\">{chromosome_dict[chromosome_to_plot]}\") and not chromosome_found:\n",
    "                continue\n",
    "            if line.startswith (f\">{chromosome_dict[chromosome_to_plot]}\") and not chromosome_found:\n",
    "                chromosome_found = True\n",
    "                continue\n",
    "            if chromosome_found:\n",
    "                if line.startswith(\">\"):\n",
    "                    break\n",
    "                else:\n",
    "                    chromosome_length = chromosome_length + len(line.strip())   \n",
    "    \n",
    "    else:\n",
    "        chromosome_length = None\n",
    "\n",
    "    return chromosome_length, chromosome_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23ee1065",
   "metadata": {
    "lines_to_next_cell": 2
   },
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "def DefineRegion(region_to_plot, chromosome_to_plot, plots_folder):\n",
    "    \"\"\"\n",
    "    Define the region to plot\n",
    "    \"\"\"\n",
    "\n",
    "    start = None\n",
    "    end = None\n",
    "\n",
    "    if region_to_plot:\n",
    "        start, end = region_to_plot.split(\"-\")\n",
    "        output_png = f\"{plots_folder}pangenome_{chromosome_to_plot}_{region_to_plot}.png\"\n",
    "        region_to_plot = f\"{start}-{end}\"\n",
    "\n",
    "    else:\n",
    "        output_png = f\"{plots_folder}pangenome_FULL_{chromosome_to_plot}.png\"\n",
    "    \n",
    "    return start, end, output_png"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c60829a",
   "metadata": {
    "lines_to_next_cell": 2
   },
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "def PlotPangenome(plots_folder, chromosome_to_plot, chromosome_length, chromosome_dict, region_to_plot, output_png, reference_name, haplotype_names):\n",
    "    \n",
    "    pangenome_haplotypes = [f for f in os.listdir(plots_folder) if f.endswith('_haplotype.bed')]\n",
    "    if reference_name in pangenome_haplotypes:\n",
    "        pangenome_haplotypes.remove(reference_name)\n",
    "\n",
    "    print(\"there are this many haplotypes in the pangenome:\")\n",
    "    print(len(pangenome_haplotypes))\n",
    "    print(pangenome_haplotypes)\n",
    "\n",
    "\n",
    "    track_ini = f\"{plots_folder}FULLpangenome_track.ini\"\n",
    "\n",
    "    if os.path.exists(track_ini):\n",
    "        print(\"track.ini already exists, deleting and make a new file...\")\n",
    "        os.remove(track_ini)\n",
    "\n",
    "    with open(track_ini, 'w') as track:\n",
    "        for file in haplotype_names:\n",
    "            track.write(f\"[haplotype_{file}]\\n\")\n",
    "            track.write(f\"file = {plots_folder}{file}_haplotype.bed\\n\")\n",
    "            print(f\"loading for ploting in following order: {plots_folder}{file}_haplotype.bed\")\n",
    "            track.write(f\"title = {file}\\n\")\n",
    "            track.write(f\"height = 0.5\\n\")\n",
    "            track.write(f\"color = bed_rgb\\n\")\n",
    "            track.write(\"display = collapsed\\n\")\n",
    "            track.write(\"labels = false\\n\")\n",
    "            track.write(\"border_color = black\\n\")\n",
    "            track.write(\"line_width = 0.05\\n\")\n",
    "    #       track.write(\"type = bed\\n\")    #by default is the extension, not necessary\n",
    "            track.write(\"\\n\")\n",
    "            track.write(\"[spacer]\\nheight = 0.5\\n\")\n",
    "\n",
    "        track.write(f\"[spacer]\\nheight = 0.5\\n\")\n",
    "        track.write(f\"[haplotype_{reference_name}]\\n\")\n",
    "        track.write(f\"file = {plots_folder}{reference_name}_haplotype.bed\\n\")\n",
    "        track.write(f\"title = Reference_{reference_name}\\n\")\n",
    "        track.write(f\"height = 1\\n\")\n",
    "        track.write(f\"color = black\\n\")\n",
    "        track.write(\"display = collapsed\\n\")\n",
    "        track.write(\"labels = false\\n\")\n",
    "        track.write(\"border_color = white\\n\")\n",
    "        track.write(\"line_width = 0.05\\n\")\n",
    "    #       track.write(\"type = bed\\n\")    #by default is the extension, not necessary\n",
    "        track.write(\"\\n\")\n",
    "        track.write(\"[spacer]\\nheight = 0.1\\n\")\n",
    "\n",
    "        track.write(\"[x-axis]\\n\")\n",
    "        track.write(\"where = bottom\\n\")\n",
    "        #track.write(\"label = true\\n\")\n",
    "        #track.write(\"font_size = 10\\n\")\n",
    "        #track.write(f\"title = Mediterranean pangenome {chromosome_to_plot} using as reference MorexV3\\n\")\n",
    "        track.write(\"[spacer]\\nheight = 0.5\\n\")\n",
    "\n",
    "    print(f\"track.ini file is {track_ini}\")\n",
    "\n",
    "    if not region_to_plot == None:\n",
    "        command_Genome_tracks = f\"--tracks {track_ini} -o {output_png} --region {chromosome_dict[chromosome_to_plot]}:{region_to_plot} --dpi 300 --width 75 --fontSize 15\"\n",
    "\n",
    "    else:   #by default whole chromosome\n",
    "        command_Genome_tracks = f\"--tracks {track_ini} -o {output_png} --region {chromosome_dict[chromosome_to_plot]}:1-{chromosome_length} --dpi 300 --width 75 --fontSize 15 \"\n",
    "\n",
    "    print(command_Genome_tracks)\n",
    "\n",
    "    subprocess.run(f\"pyGenomeTracks {command_Genome_tracks}\", shell=True)\n",
    "\n",
    "    print(output_png)\n",
    "\n",
    "    print(f\"output file is {output_png}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5eee85d5",
   "metadata": {
    "lines_to_next_cell": 2
   },
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "import os\n",
    "import gzip\n",
    "import re\n",
    "import subprocess\n",
    "import argparse\n",
    "\n",
    "def main():\n",
    "\n",
    "    if AmIaNotebook() == False:\n",
    "        parser = argparse.ArgumentParser(description=main.__doc__)\n",
    "        parser.add_argument('--hvcf-folder', \"-hvcf\", help='Folder with the haplotype VCF files', required=True)\n",
    "        parser.add_argument('--reference-hvcf', \"-ref\", help='Reference genome hvcf file', required=True)\n",
    "        parser.add_argument('--chromosome', \"-chr\", help='Chromosome to plot. Enter: chrX (chr1, chr7...)', required=True)\n",
    "        parser.add_argument('--region', \"-reg\", help='Region to . Add it with a \"-\" dividing the start-end (100000-245000). If no value is provided, by default whole chr will be ploted', required=False)\n",
    "        parser.add_argument('--reference-fasta', \"-fa\", help='Reference genome fasta file', required=True)\n",
    "\n",
    "        args = parser.parse_args()\n",
    "\n",
    "        hvcf_folder = args.hvcf_folder\n",
    "        reference_hvcf = args.reference_hvcf\n",
    "        chromosome_to_plot = args.chromosome\n",
    "        region_to_plot = args.region\n",
    "        ref_fasta = args.reference_fasta\n",
    "\n",
    "    else:    \n",
    "\n",
    "        hvcf_folder = \"/genoma/nfs/PHG/output/vcf_files/\"\n",
    "        chromosome_to_plot = \"chr7\" #give imput of which chromosome to plot\n",
    "        reference_hvcf = \"/genoma/nfs/PHG/vcf_dbs/hvcf_files/MorexV3.h.vcf.gz\"\n",
    "        ref_fasta = \"/genoma/nfs/PHG/data/prepared_genomes/MorexV3.fa\"\n",
    "        region_to_plot = None\n",
    "\n",
    "    if not region_to_plot:\n",
    "        region_to_plot = None\n",
    "\n",
    "    plots_folder = f\"{hvcf_folder}plots/\"\n",
    "\n",
    "    haplotype_files, haplotype_names, reference_name = DefineHVCFs(hvcf_folder, reference_hvcf)\n",
    "\n",
    "    colors_dic, ref_color, columns_header = PangenomeColors(reference_name, haplotype_names)\n",
    "\n",
    "    CreateBedFile(hvcf_folder, plots_folder, haplotype_names, reference_name, reference_hvcf, colors_dic, columns_header)\n",
    "\n",
    "    chromosome_length, chromosome_dict = WholeChrLenght(ref_fasta, chromosome_to_plot, reference_name, region_to_plot)\n",
    "\n",
    "    start, end, output_png = DefineRegion(region_to_plot, chromosome_to_plot, plots_folder)\n",
    "\n",
    "    PlotPangenome(plots_folder, chromosome_to_plot, chromosome_length, chromosome_dict, region_to_plot, output_png, reference_name, haplotype_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68c26e60",
   "metadata": {
    "lines_to_next_cell": 2
   },
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    try:\n",
    "        main()    \n",
    "    except KeyboardInterrupt:\n",
    "        print(\"Script interrupted by user. Exiting...\")\n",
    "        sys.exit(0)\n",
    "        raise KeyboardInterrupt"
   ]
  }
 ],
 "metadata": {
  "jupytext": {
   "cell_metadata_filter": "-all",
   "encoding": "# coding: utf-8",
   "executable": "/usr/bin/env python",
   "main_language": "python",
   "notebook_metadata_filter": "-all"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
