{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "c8efc3bb",
   "metadata": {},
   "source": [
    "\n",
    "# Script to start from the hVCF of the reference, and then add (merging) one by one the others hvcf of the pangenome, and see how many ranges does it increase.\n",
    "# \n",
    "# Takes advantage of the phg merge function, in which each file has a key for the range of the pangenome, if its exists, using the reference coordenades.\n",
    "# You can compare each range for the genomes having each range:\n",
    "# \n",
    "# \n",
    "# \n",
    "# \n",
    "cat output/vcf_files/merged_hvcfs_19092024.h.vcf | grep -v \"##ALT\"      #picking a random range\n",
    "# \n",
    "# chr3H_LR890098.1        613549767       .       T       <bd32348136f0502d9356256f7543dc9c>,<78ffc528bd70012d1f75067ecbd7f4e8>   .       .       END=613553517   .............................\n",
    "# \n",
    "# ## These are all different ranges that fits in this range, doing a quick search:\n",
    "# \n",
    "# line: HOR_13942\n",
    "# chr3H_LR890098.1\tbd32348136f0502d9356256f7543dc9c\tSTART=613549767\tEND=613553517\n",
    "# lenght= 3750 bp\n",
    "# \n",
    "# line: HOR_12184\n",
    "# line: HOR_1168\n",
    "# chr3H_LR890098.1\tbd32348136f0502d9356256f7543dc9c\tSTART=613549767\tEND=613553517\n",
    "# lenght= 3750 bp\n",
    "# \n",
    "# line: HOR_2830\n",
    "# chr3H_LR890098.1\tbd32348136f0502d9356256f7543dc9c\tSTART=613549767\tEND=613553517\n",
    "# lenght= 3750 bp\n",
    "# \n",
    "# line: HOR_21599\n",
    "# chr3H_LR890098.1\tbd32348136f0502d9356256f7543dc9c\tSTART=613549767\tEND=613553517\n",
    "# lenght= 3750 bp\n",
    "# \n",
    "# line: HOR_14121\n",
    "# chr3H_LR890098.1\tbd32348136f0502d9356256f7543dc9c\tSTART=613549767\tEND=613553517\n",
    "# lenght= 3750 bp\n",
    "# \n",
    "# line: HOR_21595\n",
    "# chr3H_LR890098.1\tbd32348136f0502d9356256f7543dc9c\tSTART=613549767\tEND=613553517\n",
    "# lenght= 3750 bp\n",
    "# \n",
    "# line: HOR_10892\n",
    "# chr3H_LR890098.1\tbd32348136f0502d9356256f7543dc9c\tSTART=613549767\tEND=613553517\n",
    "# lenght= 3750 bp\n",
    "# \n",
    "# line: HOR_2779\n",
    "# chr3H_LR890098.1\tbd32348136f0502d9356256f7543dc9c\tSTART=613549767\tEND=613553517\n",
    "# lenght= 3750 bp\n",
    "# \n",
    "# line: HOR_3474\n",
    "# chr3H_LR890098.1\tbd32348136f0502d9356256f7543dc9c\tSTART=613549767\tEND=613553517\n",
    "# lenght= 3750 bp\n",
    "# \n",
    "# line: HOR_3365\n",
    "# chr3H_LR890098.1\tbd32348136f0502d9356256f7543dc9c\tSTART=613549767\tEND=613553517\n",
    "# lenght= 3750 bp\n",
    "# \n",
    "# line: HOR_12184\n",
    "# chr3H_LR890098.1\t78ffc528bd70012d1f75067ecbd7f4e8\tSTART=613549767\tEND=613553517\n",
    "# \n",
    "# ## That means:\n",
    "# 1. There are 10 haplotypes which have been stablish as conserved. \n",
    "# 2. There is 1 haplotype which has been determined as different\n",
    "# 3. Any other variety having not the range, is because is not present\n",
    "# \n",
    "# --> We can detect presence/absence/variation of ranges.\n",
    "# \n",
    "#  ## How to work\n",
    "# \n",
    "#  1. Compare the number of ranges (lines) at the hVCF when having 1 genome, then 2, then 3...\n",
    "#  This confirms each genome is adding new genes, at least one per each range (atm we dont know how many, or the absence of them at new included genomes)\n",
    "#  2. Compare the number of range keys. An increasing of nÂº of keys means an introduction of variation in existing genes. (kind of \"alleles of ranges\")\n",
    "# \n",
    "# \n",
    "# ** Works with phgv2-conda env\n",
    "# "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5329e21f",
   "metadata": {
    "lines_to_next_cell": 2
   },
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "def AmIaNotebook():\n",
    "    try:\n",
    "        shell = get_ipython().__class__.__name__\n",
    "        if shell == 'ZMQInteractiveShell':\n",
    "            return True  # Jupyter notebook or qtconsole\n",
    "            print('Jupyter notebook or qtconsole')\n",
    "        elif shell == 'TerminalInteractiveShell':\n",
    "            return False  # Terminal running IPython\n",
    "            print('Terminal running')\n",
    "        else:\n",
    "            return False  # Other type of environment\n",
    "            print('Other type of environment')\n",
    "    except NameError:\n",
    "        return False  # Not in an interactive environment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "883c20d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "def import_pangenome_genomes(hvcf_folder):\n",
    "\n",
    "    hvcf_files = glob.glob(hvcf_folder + '/*.h.vcf.gz')\n",
    "\n",
    "    print (f\"There are {len(hvcf_files)} genomes at the pangenome\")\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "4779cb7f",
   "metadata": {
    "lines_to_next_cell": 2
   },
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "#open the temporary folder\n",
    "\n",
    "def open_temp_folder_RangePangenomeEvolution(hvcf_folder):\n",
    "    hvcf_files = glob.glob(hvcf_folder + '/*.h.vcf.gz')\n",
    "\n",
    "    temp_folder = f\"{hvcf_folder}/temp\"\n",
    "\n",
    "    if os.path.exists(temp_folder):\n",
    "        print(\"The temporary folder already exists, deleting it and whole content and opening it again\")\n",
    "        shutil.rmtree(temp_folder)\n",
    "        os.mkdir(temp_folder)\n",
    "    else:\n",
    "        os.mkdir(temp_folder)\n",
    "        print(\"The temporary folder has been created\")\n",
    "\n",
    "    return temp_folder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "2268c768",
   "metadata": {
    "lines_to_next_cell": 2
   },
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "def loophvcf_files(hvcf_folder, reference_file, range_bedfile):\n",
    "    hvcf_files = glob.glob(hvcf_folder + '/*.h.vcf.gz')\n",
    "    temp_folder = f\"{hvcf_folder}/temp/\"\n",
    "\n",
    "    hvcf_processed = []\n",
    "    genomes_count = 0\n",
    "\n",
    "    for file in hvcf_files:\n",
    "        print (f\"Processing {file}\")\n",
    "        os.system(f\"cp {file} {temp_folder}\")\n",
    "        genomes_count += 1\n",
    "        hvcf_processed.append(file)\n",
    "\n",
    "        command = f\"phg merge-hvcfs --input-dir {temp_folder} --output-file {temp_folder}{genomes_count}.h.vcf --id-format CHECKSUM --reference-file {reference_file} --range-bedfile {range_bedfile}\"\n",
    "        print(command)  \n",
    "        os.system(command)\n",
    "\n",
    "        #remove the copied file and camuflate the generated files to not be used or deleted. Tricking the first one to be conservated and used as well.\n",
    "        os.system(f\"rm {temp_folder}*.h.vcf.gz\")\n",
    "\n",
    "        if genomes_count > 1:\n",
    "            os.system(f\"mv {temp_folder}{genomes_count - 1}.h.vcf {temp_folder}{genomes_count - 1}.h.vcf.YouCanNotSeeMe\")\n",
    "        else:\n",
    "            #if the file exists\n",
    "            if os.path.exists(f\"{temp_folder}{genomes_count - 2}.h.vcf\"):\n",
    "                os.system(f\"mv {temp_folder}{genomes_count - 2}.h.vcf {temp_folder}{genomes_count - 2}.h.vcf.YouCanNotSeeMe \")\n",
    "\n",
    "    #Change also the last file out of the loop, since is not gonna be used\n",
    "    os.system(f\"mv {temp_folder}{genomes_count}.h.vcf {temp_folder}{genomes_count}.h.vcf.YouCanNotSeeMe\") "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e00ca54a",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "def RangesAmplificationSlope (hvcf_folder):\n",
    "    \n",
    "    hvcf_files = glob.glob(hvcf_folder + '/*.h.vcf.gz')\n",
    "\n",
    "    temp_folder = f\"{hvcf_folder}temp/\"\n",
    "    temp_files = glob.glob(temp_folder + '/*.h.vcf.YouCanNotSeeMe')\n",
    "\n",
    "    #sort temporarly files, by number from 0 to n\n",
    "    temp_files.sort(key = lambda x: int(x.split(\"/\")[-1].split(\".\")[0]))\n",
    "    #print (temp_files)\n",
    "\n",
    "    if len(temp_files) == 0:\n",
    "        raise Exception(\"There are no files to process\")\n",
    "    \n",
    "    if len(temp_files) != len (hvcf_files):\n",
    "        raise Exception(\"The number of files to process is not the same as the number of files in the pangenome\")\n",
    "\n",
    "        \n",
    "    files_processed = 0\n",
    "    ranges_dict = {}\n",
    "\n",
    "    for file in temp_files:\n",
    "        files_processed += 1\n",
    "        #print (f\"Processing {files_processed} genomes\")\n",
    "        command = f'cat {temp_folder}{files_processed}.h.vcf.YouCanNotSeeMe | grep -v \"#\" | wc -l'\n",
    "        range = int(subprocess.check_output(command, shell=True).decode().strip())\n",
    "        print (f\"Genomes processed: {files_processed} - Range: {range}\")\n",
    "\n",
    "        #save in a dictionary the number of files procesed and the number of range\n",
    "        ranges_dict[files_processed] = range\n",
    "\n",
    "    #print (ranges_dict)\n",
    "    return (ranges_dict)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "0f8555b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "def PlotRangesSlope (hvcf_folder):\n",
    "\n",
    "    ranges_dict = RangesAmplificationSlope(hvcf_folder)\n",
    "\n",
    "    #plot the ranges\n",
    "    plt.plot(ranges_dict.keys(), ranges_dict.values())\n",
    "    plt.xlabel('Genomes')\n",
    "    plt.ylabel('Ranges')\n",
    "    plt.title('Ranges Amplification Slope')\n",
    "    plt.savefig(f\"{hvcf_folder}/RangesAmplificationSlope.png\")\n",
    "    plt.show()\n",
    "\n",
    "\n",
    "# Until here the code works on the number of ranges. \n",
    "# Now lets take advantage of the files generated and lets call for variation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "7bd75631",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "def RangesVariationSlope (hvcf_folder):\n",
    "\n",
    "    temp_folder = f\"{hvcf_folder}temp/\"\n",
    "    temp_files = glob.glob(temp_folder + '/*.h.vcf.YouCanNotSeeMe')\n",
    "    temp_files.sort(key = lambda x: int(x.split(\"/\")[-1].split(\".\")[0]))\n",
    "    if len(temp_files) == 0:\n",
    "        raise Exception(\"There are no files to process\")\n",
    "\n",
    "    ranges_dict = RangesAmplificationSlope(hvcf_folder) \n",
    "\n",
    "    files_processed = 0\n",
    "    keys_dict = {}\n",
    "\n",
    "    for file in temp_files:\n",
    "        files_processed += 1\n",
    "        #print (f\"Processing {files_processed} genomes\")\n",
    "        command = f'cat {temp_folder}{files_processed}.h.vcf.YouCanNotSeeMe | grep \"##ALT\" | wc -l'\n",
    "        alleles = int(subprocess.check_output(command, shell=True).decode().strip())\n",
    "        print (f\"Genome {files_processed} has {alleles} alleles\")\n",
    "        #print(alleles)\n",
    "\n",
    "        \n",
    "    #Save in a dictionary\n",
    "        keys_dict[files_processed] = alleles\n",
    "\n",
    "    files_processed = 0\n",
    "    swap_rate = {}\n",
    "    for file in temp_files:\n",
    "        files_processed += 1\n",
    "        if files_processed == 1:\n",
    "            swap_rate[file] = None\n",
    "        else:\n",
    "            swap_rate[file] = keys_dict[files_processed] / keys_dict[files_processed- 1]\n",
    "            print (f\"Swap rate between {files_processed} and {files_processed - 1} is {swap_rate[file]}\")\n",
    "    \n",
    "    return (swap_rate)\n",
    "\n",
    "    \n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ae0f4c29",
   "metadata": {
    "lines_to_next_cell": 2
   },
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "def PlotAllelesSlope(hvcf_folder):\n",
    "    keys_dict = RangesVariationSlope(hvcf_folder)\n",
    "\n",
    "    plt.plot(keys_dict.keys(), keys_dict.values())\n",
    "    plt.xlabel('Genomes')\n",
    "    plt.ylabel('Ranges \"alleles\"')\n",
    "    plt.title('Ranges Variation Slope')\n",
    "    #plt.savefig(f\"{hvcf_folder}/RangesVariationSlope.png\")\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5c0b5abf",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "5cbf1a68",
   "metadata": {
    "lines_to_next_cell": 2
   },
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "import glob\n",
    "import os\n",
    "import glob\n",
    "import shutil\n",
    "import sys\n",
    "import subprocess\n",
    "import matplotlib.pyplot as plt\n",
    "import inspect\n",
    "import argparse\n",
    "from IPython import get_ipython\n",
    "\n",
    "def main():\n",
    "    def AmIaNotebook():\n",
    "        try:\n",
    "            shell = get_ipython().__class__.__name__\n",
    "            if shell == 'ZMQInteractiveShell':\n",
    "                return True  # Jupyter notebook or qtconsole\n",
    "            elif shell == 'TerminalInteractiveShell':\n",
    "                return False  # Terminal running IPython\n",
    "            else:\n",
    "                return False  # Other type of environment\n",
    "        except NameError:\n",
    "            return False  # Not in an interactive environment\n",
    "\n",
    "    #Check if the name of this file ends with .py or with .\n",
    "\n",
    "    if AmIaNotebook() == False:\n",
    "        parser = argparse.ArgumentParser(description=main.__doc__)\n",
    "        parser.add_argument('--hvcf-folder', \"-hf\", type=str, help='The folder with the hvcf files')\n",
    "        parser.add_argument('--reference-file', \"-ref\", type=str, help='The reference file')\n",
    "        parser.add_argument('--range-bedfile', \"-bed\", type=str, help='The range bedfile')\n",
    "        args = parser.parse_args()\n",
    "\n",
    "        hvcf_folder = args.hvcf_folder\n",
    "        reference_file = args.reference_file\n",
    "        range_bedfile = args.range_bedfile\n",
    "\n",
    "    else:\n",
    "        \n",
    "        hvcf_folder = \"/scratch/PHG/output/vcf_files/\"\n",
    "        reference_file = \"/scratch/PHG/data/prepared_genomes/MorexV3.fa\"\n",
    "        range_bedfile = \"/scratch/PHG/data/ref_ranges.bed\"\n",
    "\n",
    "\n",
    "    temp_folder = f\"{hvcf_folder}/temp\"\n",
    "\n",
    "    import_pangenome_genomes(hvcf_folder)\n",
    "\n",
    "    open_temp_folder_RangePangenomeEvolution(hvcf_folder)  #Commented to avoid deleting the folder every single time\n",
    "\n",
    "    loophvcf_files(hvcf_folder, reference_file, range_bedfile)     #Lo comento para no repetir durante los tests\n",
    "\n",
    "    RangesAmplificationSlope(hvcf_folder)\n",
    "\n",
    "    PlotRangesSlope(hvcf_folder)\n",
    "\n",
    "    RangesVariationSlope(hvcf_folder)\n",
    "\n",
    "    PlotAllelesSlope(hvcf_folder)\n",
    "\n",
    "    #At some point, delete whole temp folder\n",
    "    shutil.rmtree(temp_folder)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "9d3396c2",
   "metadata": {
    "lines_to_next_cell": 2
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 11 genomes at the pangenome\n",
      "The temporary folder has been created\n",
      "Processing /scratch/PHG/output/vcf_files/HOR_14121.h.vcf.gz\n",
      "phg merge-hvcfs --input-dir /scratch/PHG/output/vcf_files//temp/ --output-file /scratch/PHG/output/vcf_files//temp/1.h.vcf --id-format CHECKSUM --reference-file /scratch/PHG/data/prepared_genomes/MorexV3.fa --range-bedfile /scratch/PHG/data/ref_ranges.bed\n",
      "[DefaultDispatcher-worker-1] INFO net.maizegenetics.phgv2.api.HaplotypeGraph 2024-10-02 16:42:59,244: processFiles: /scratch/PHG/output/vcf_files/temp/HOR_14121.h.vcf.gz\n",
      "[main] INFO net.maizegenetics.phgv2.api.HaplotypeGraph 2024-10-02 16:43:01,427: rangeToSampleToChecksum: 62489 x 1\n",
      "[main] INFO net.maizegenetics.phgv2.api.HaplotypeGraph 2024-10-02 16:43:01,428: numOfSamples: 1\n",
      "[main] INFO net.maizegenetics.phgv2.api.HaplotypeGraph 2024-10-02 16:43:01,428: numOfRanges: 62489\n",
      "Processing /scratch/PHG/output/vcf_files/HOR_2830.h.vcf.gz\n",
      "phg merge-hvcfs --input-dir /scratch/PHG/output/vcf_files//temp/ --output-file /scratch/PHG/output/vcf_files//temp/2.h.vcf --id-format CHECKSUM --reference-file /scratch/PHG/data/prepared_genomes/MorexV3.fa --range-bedfile /scratch/PHG/data/ref_ranges.bed\n",
      "[DefaultDispatcher-worker-1] INFO net.maizegenetics.phgv2.api.HaplotypeGraph 2024-10-02 16:43:24,418: processFiles: /scratch/PHG/output/vcf_files/temp/1.h.vcf\n",
      "[DefaultDispatcher-worker-1] INFO net.maizegenetics.phgv2.api.HaplotypeGraph 2024-10-02 16:43:25,810: processFiles: /scratch/PHG/output/vcf_files/temp/HOR_2830.h.vcf.gz\n",
      "[main] INFO net.maizegenetics.phgv2.api.HaplotypeGraph 2024-10-02 16:43:28,198: rangeToSampleToChecksum: 63744 x 2\n",
      "[main] INFO net.maizegenetics.phgv2.api.HaplotypeGraph 2024-10-02 16:43:28,199: numOfSamples: 2\n",
      "[main] INFO net.maizegenetics.phgv2.api.HaplotypeGraph 2024-10-02 16:43:28,199: numOfRanges: 63744\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    try:\n",
    "        main()\n",
    "    except KeyboardInterrupt:\n",
    "        print(\"Script interrupted by user. Exiting...\")\n",
    "        shutil.rmtree(temp_folder)\n",
    "        sys.exit(0)\n",
    "        raise KeyboardInterrupt\n",
    "        \n"
   ]
  }
 ],
 "metadata": {
  "jupytext": {
   "cell_metadata_filter": "-all",
   "encoding": "# coding: utf-8",
   "executable": "/usr/bin/env python",
   "main_language": "python",
   "notebook_metadata_filter": "-all"
  },
  "kernelspec": {
   "display_name": "phgv2-conda",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
